File: src/test/java/com/github/nsc/de/compiler/interpreter/values/IntegerValueTest.java
package com.github.nsc.de.compiler.interpreter.values;

import org.junit.jupiter.api.Test;

import static com.github.nsc.de.compiler.interpreter.values.IntegerValue.*;
import static com.github.nsc.de.compiler.interpreter.values.BooleanValue.*;
import static org.junit.jupiter.api.Assertions.*;

public class IntegerValueTest {

    @Test
    public void testAdd() {
        assertEquals(2, ((IntegerValue) ONE.add(ONE)).getValue());
        assertEquals(0, ((IntegerValue) new IntegerValue(100).add(new IntegerValue(-100))).getValue());
    }

    @Test
    public void testSub() {
        assertEquals(0, ((IntegerValue) ONE.sub(ONE)).getValue());
        assertEquals(200, ((IntegerValue) new IntegerValue(100).sub(new IntegerValue(-100))).getValue());
    }

    @Test
    public void testMul() {
        assertEquals(1, ((IntegerValue) ONE.mul(ONE)).getValue());
        assertEquals(-10_000, ((IntegerValue) new IntegerValue(100).mul(new IntegerValue(-100))).getValue());
    }

    @Test
    public void testDiv() {
        assertEquals(5, ((IntegerValue) new IntegerValue(10).div(new IntegerValue(2))).getValue());
        assertEquals(6, ((IntegerValue) new IntegerValue(48).div(new IntegerValue(7))).getValue());
        assertEquals(2.2, ((DoubleValue) new IntegerValue(11).div(new DoubleValue(5))).getValue());
    }

    @Test
    public void testPow() {
        assertEquals(4, ((IntegerValue) new IntegerValue(2).pow(new IntegerValue(2))).getValue());
        assertEquals(0, ((IntegerValue) new IntegerValue(5).pow(new IntegerValue(-2))).getValue());
        assertEquals(2, ((DoubleValue) new IntegerValue(4).pow(new DoubleValue(.5))).getValue());
    }



    @Test
    public void testOr() {
        Error error = assertThrows(Error.class, () -> ONE.or(ONE));
        assertEquals("Operator '||' is not defined for type integer", error.getMessage());
    }

    @Test
    public void testAnd() {
        Error error = assertThrows(Error.class, () -> ONE.and(ONE));
        assertEquals("Operator '&&' is not defined for type integer", error.getMessage());
    }



    @Test
    public void testEqualsEquals() {
        assertEquals(TRUE, ONE.equals(ONE));
        assertEquals(TRUE, ONE.equals(new DoubleValue(1.0)));
        assertEquals(TRUE, ONE.equals(new IntegerValue(1)));
        assertEquals(FALSE, ONE.equals(new IntegerValue(2)));
        assertEquals(FALSE, ONE.equals(new DoubleValue(1.1)));
    }

    @Test
    public void testBiggerEquals() {
        assertEquals(TRUE, ONE.bigger_equals(ONE));
        assertEquals(TRUE, ONE.bigger_equals(new DoubleValue(1.0)));
        assertEquals(TRUE, ONE.bigger_equals(new IntegerValue(1)));
        assertEquals(TRUE, ONE.bigger_equals(new IntegerValue(-10)));
        assertEquals(TRUE, ONE.bigger_equals(new DoubleValue(-10.0)));
        assertEquals(FALSE, ONE.bigger_equals(new IntegerValue(10)));
        assertEquals(FALSE, ONE.bigger_equals(new DoubleValue(10.0)));
    }

    @Test
    public void testSmallerEquals() {
        assertEquals(TRUE, ONE.smaller_equals(ONE));
        assertEquals(TRUE, ONE.smaller_equals(new DoubleValue(1.0)));
        assertEquals(TRUE, ONE.smaller_equals(new IntegerValue(1)));
        assertEquals(TRUE, ONE.smaller_equals(new IntegerValue(10)));
        assertEquals(TRUE, ONE.smaller_equals(new DoubleValue(10.0)));
        assertEquals(FALSE, ONE.smaller_equals(new IntegerValue(-10)));
        assertEquals(FALSE, ONE.smaller_equals(new DoubleValue(-10.0)));
    }

    @Test
    public void testBigger() {
        assertEquals(FALSE, ONE.bigger(ONE));
        assertEquals(FALSE, ONE.bigger(new DoubleValue(1.0)));
        assertEquals(FALSE, ONE.bigger(new IntegerValue(1)));
        assertEquals(TRUE, ONE.bigger(new IntegerValue(-10)));
        assertEquals(TRUE, ONE.bigger(new DoubleValue(-10.0)));
        assertEquals(FALSE, ONE.bigger(new IntegerValue(10)));
        assertEquals(FALSE, ONE.bigger(new DoubleValue(10.0)));
    }

    @Test
    public void testSmaller() {
        assertEquals(FALSE, ONE.smaller(ONE));
        assertEquals(FALSE, ONE.smaller(new DoubleValue(1.0)));
        assertEquals(FALSE, ONE.smaller(new IntegerValue(1)));
        assertEquals(TRUE, ONE.smaller(new IntegerValue(10)));
        assertEquals(TRUE, ONE.smaller(new DoubleValue(10.0)));
        assertEquals(FALSE, ONE.smaller(new IntegerValue(-10)));
        assertEquals(FALSE, ONE.smaller(new DoubleValue(-10.0)));
    }



}

File: src/test/kotlin/com/github/nsc/de/shake/interpreter/values/IntegerValueTest.kt
package com.github.nsc.de.shake.interpreter.values

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class IntegerValueTest {
    @Test
    fun testAdd() {
        Assertions.assertEquals(2, (IntegerValue.ONE.add(IntegerValue.ONE) as IntegerValue).value)
        Assertions.assertEquals(0, (IntegerValue(100).add(IntegerValue(-100)) as IntegerValue).value)
        Assertions.assertEquals(2.0, (IntegerValue.ONE.add(DoubleValue(1.0)) as DoubleValue).value)
        Assertions.assertEquals(0.0, (IntegerValue(100).add(DoubleValue(-100.0)) as DoubleValue).value)
    }

    @Test
    fun testSub() {
        Assertions.assertEquals(0, (IntegerValue.ONE.sub(IntegerValue.ONE) as IntegerValue).value)
        Assertions.assertEquals(200, (IntegerValue(100).sub(IntegerValue(-100)) as IntegerValue).value)
        Assertions.assertEquals(0.0, (IntegerValue.ONE.sub(DoubleValue(1.0)) as DoubleValue).value)
        Assertions.assertEquals(200.0, (IntegerValue(100).sub(DoubleValue(-100.0)) as DoubleValue).value)
    }

    @Test
    fun testMul() {
        Assertions.assertEquals(1, (IntegerValue.ONE.mul(IntegerValue.ONE) as IntegerValue).value)
        Assertions.assertEquals(-10000, (IntegerValue(100).mul(IntegerValue(-100)) as IntegerValue).value)
        Assertions.assertEquals(1.0, (IntegerValue.ONE.mul(DoubleValue(1.0)) as DoubleValue).value)
        Assertions.assertEquals(-10000.0, (IntegerValue(100).mul(DoubleValue(-100.0)) as DoubleValue).value)
    }

    @Test
    fun testDiv() {
        Assertions.assertEquals(5, (IntegerValue(10).div(IntegerValue(2)) as IntegerValue).value)
        Assertions.assertEquals(6, (IntegerValue(48).div(IntegerValue(7)) as IntegerValue).value)
        Assertions.assertEquals(2.2, (IntegerValue(11).div(DoubleValue(5.0)) as DoubleValue).value)
    }

    @Test
    fun testMod() {
        Assertions.assertEquals(0, (IntegerValue(10).mod(IntegerValue(2)) as IntegerValue).value)
        Assertions.assertEquals(6, (IntegerValue(48).mod(IntegerValue(7)) as IntegerValue).value)
        Assertions.assertEquals(1.0, (IntegerValue(11).mod(DoubleValue(5.0)) as DoubleValue).value)
    }

    @Test
    fun testPow() {
        Assertions.assertEquals(4, (IntegerValue(2).pow(IntegerValue(2)) as IntegerValue).value)
        Assertions.assertEquals(0, (IntegerValue(5).pow(IntegerValue(-2)) as IntegerValue).value)
        Assertions.assertEquals(2.0, (IntegerValue(4).pow(DoubleValue(.5)) as DoubleValue).value)
    }

    @Test
    fun testOr() {
        val error = Assertions.assertThrows(Error::class.java) { IntegerValue.ONE.or(IntegerValue.ONE) }
        Assertions.assertEquals("Operator '||' is not defined for type integer", error.message)
    }

    @Test
    fun testAnd() {
        val error = Assertions.assertThrows(Error::class.java) { IntegerValue.ONE.and(IntegerValue.ONE) }
        Assertions.assertEquals("Operator '&&' is not defined for type integer", error.message)
    }

    @Test
    fun testEqualsEquals() {
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.equals(IntegerValue.ONE))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.equals(DoubleValue(1.0)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.equals(IntegerValue(1)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.equals(IntegerValue(2)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.equals(DoubleValue(1.1)))
    }

    @Test
    fun testBiggerEquals() {
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.bigger_equals(IntegerValue.ONE))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.bigger_equals(DoubleValue(1.0)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.bigger_equals(IntegerValue(1)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.bigger_equals(IntegerValue(-10)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.bigger_equals(DoubleValue(-10.0)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.bigger_equals(IntegerValue(10)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.bigger_equals(DoubleValue(10.0)))
    }

    @Test
    fun testSmallerEquals() {
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.smaller_equals(IntegerValue.ONE))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.smaller_equals(DoubleValue(1.0)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.smaller_equals(IntegerValue(1)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.smaller_equals(IntegerValue(10)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.smaller_equals(DoubleValue(10.0)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.smaller_equals(IntegerValue(-10)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.smaller_equals(DoubleValue(-10.0)))
    }

    @Test
    fun testBigger() {
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.bigger(IntegerValue.ONE))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.bigger(DoubleValue(1.0)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.bigger(IntegerValue(1)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.bigger(IntegerValue(-10)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.bigger(DoubleValue(-10.0)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.bigger(IntegerValue(10)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.bigger(DoubleValue(10.0)))
    }

    @Test
    fun testSmaller() {
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.smaller(IntegerValue.ONE))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.smaller(DoubleValue(1.0)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.smaller(IntegerValue(1)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.smaller(IntegerValue(10)))
        Assertions.assertEquals(BooleanValue.TRUE, IntegerValue.ONE.smaller(DoubleValue(10.0)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.smaller(IntegerValue(-10)))
        Assertions.assertEquals(BooleanValue.FALSE, IntegerValue.ONE.smaller(DoubleValue(-10.0)))
    }
}